File: packages/protocol-ts/src/provenance.ts
import type { ResearchProvenanceFingerprint } from "./generated/loop/v1/research_common_pb.js";

export const PROVENANCE_COMPONENTS = Object.freeze([
  "source_code",
  "operator_registry",
  "configuration",
  "data_manifest",
  "trading_calendar",
  "environment",
] as const);

export type ProvenanceComponent = (typeof PROVENANCE_COMPONENTS)[number];

export type ProvenanceErrorCode =
  | "invalid_digest"
  | "recording_mismatch"
  | "stale"
  | "unresolved_current";

/** Integrity/freshness failure, never deterministic factor rejection. */
export class ProvenanceError extends Error {
  readonly changed: readonly ProvenanceComponent[];

  constructor(
    readonly code: ProvenanceErrorCode,
    changed: readonly ProvenanceComponent[] = [],
  ) {
    super(`research provenance failed: ${code}: ${changed.join(",")}`);
    this.name = "ProvenanceError";
    this.changed = Object.freeze([...changed]);
  }
}

/** Validated, immutable copies; later DTO mutation cannot alter this snapshot. */
export class ProvenanceSnapshot {
  private readonly digests: readonly string[];

  private constructor(digests: string[]) {
    this.digests = Object.freeze(digests);
    Object.freeze(this);
  }

  static fromWire(value: ResearchProvenanceFingerprint): ProvenanceSnapshot {
    const fields = [
      value.sourceCodeSha256,
      value.operatorRegistrySha256,
      value.configurationSha256,
      value.dataManifestSha256,
      value.tradingCalendarSha256,
      value.environmentSha256,
    ];
    const digests = PROVENANCE_COMPONENTS.map((component, index) => {
      const bytes = fields[index]?.value;
      if (!(bytes instanceof Uint8Array) || bytes.byteLength !== 32) {
        throw new ProvenanceError("invalid_digest", [component]);
      }
      return Array.from(bytes, (byte) => byte.toString(16).padStart(2, "0")).join("");
    });
    return new ProvenanceSnapshot(digests);
  }

  /** Every changed field in stable protocol order. */
  differences(other: ProvenanceSnapshot): readonly ProvenanceComponent[] {
    return Object.freeze(
      PROVENANCE_COMPONENTS.filter((_, index) => this.digests[index] !== other.digests[index]),
    );
  }
}

/** Metadata freshness only, not a data capability or factor admission. */
export type ProvenanceAssessment =
  | { readonly status: "current"; readonly changed: readonly [] }
  | { readonly status: "stale"; readonly changed: readonly ProvenanceComponent[] }
  | { readonly status: "unresolved"; readonly changed: readonly [] };

/**
 * Verify original-run integrity before freshness; never rewrite historical evidence.
 * The owner must resolve factor, backtest, sample, seed, frozen inputs and current
 * context independently. Matching caller metadata is not authority or execution proof.
 */
export function assessProvenance(
  recorded: ProvenanceSnapshot,
  frozen: ProvenanceSnapshot,
  current: ProvenanceSnapshot | undefined,
): ProvenanceAssessment {
  const mismatch = recorded.differences(frozen);
  if (mismatch.length !== 0) throw new ProvenanceError("recording_mismatch", mismatch);
  if (current === undefined) {
    return Object.freeze({ status: "unresolved", changed: Object.freeze([] as const) });
  }
  const changed = recorded.differences(current);
  return changed.length === 0
    ? Object.freeze({ status: "current", changed: Object.freeze([] as const) })
    : Object.freeze({ status: "stale", changed });
}

/** Refuse stale/unresolved metrics at a current-result consumption boundary. */
export function requireCurrentProvenance(assessment: ProvenanceAssessment): void {
  if (assessment.status === "stale") throw new ProvenanceError("stale", assessment.changed);
  if (assessment.status !== "current") throw new ProvenanceError("unresolved_current");
}

File: crates/loop-protocol/src/provenance.rs
//! Immutable research fingerprints and fail-closed metric freshness checks.
//!
//! Frozen and current inputs must be resolved by the owning service, not copied
//! from caller metadata. Equality is not authorization or proof of execution.
#![deny(missing_docs)]

use std::fmt;

use crate::wire::v1::ResearchProvenanceFingerprint;

/// One independently identity-bearing input, in stable protocol field order.
#[derive(Clone, Copy, Debug, Eq, PartialEq)]
pub enum ProvenanceComponent {
    /// Executable research source manifest.
    SourceCode,
    /// Versioned numerical operator definitions.
    OperatorRegistry,
    /// Resolved research configuration and policies.
    Configuration,
    /// Immutable input-data manifest.
    DataManifest,
    /// Exact exchange-session calendar.
    TradingCalendar,
    /// Resolved execution environment manifest.
    Environment,
}

impl ProvenanceComponent {
    /// Components in the order declared by `ResearchProvenanceFingerprint`.
    pub const ALL: [Self; 6] = [
        Self::SourceCode,
        Self::OperatorRegistry,
        Self::Configuration,
        Self::DataManifest,
        Self::TradingCalendar,
        Self::Environment,
    ];

    /// Stable, language-neutral field name, without the digest suffix.
    pub const fn as_str(self) -> &'static str {
        match self {
            Self::SourceCode => "source_code",
            Self::OperatorRegistry => "operator_registry",
            Self::Configuration => "configuration",
            Self::DataManifest => "data_manifest",
            Self::TradingCalendar => "trading_calendar",
            Self::Environment => "environment",
        }
    }
}

/// Validated owned bytes; later mutation of the wire DTO cannot change this value.
#[derive(Clone, Copy, Debug, Eq, PartialEq)]
pub struct ProvenanceSnapshot {
    digests: [[u8; 32]; 6],
}

impl TryFrom<&ResearchProvenanceFingerprint> for ProvenanceSnapshot {
    type Error = ProvenanceError;

    fn try_from(value: &ResearchProvenanceFingerprint) -> Result<Self, Self::Error> {
        let fields = [
            value.source_code_sha256.as_ref(),
            value.operator_registry_sha256.as_ref(),
            value.configuration_sha256.as_ref(),
            value.data_manifest_sha256.as_ref(),
            value.trading_calendar_sha256.as_ref(),
            value.environment_sha256.as_ref(),
        ];
        let mut digests = [[0; 32]; 6];
        for (index, (component, field)) in
            ProvenanceComponent::ALL.into_iter().zip(fields).enumerate()
        {
            digests[index] = field
                .and_then(|digest| digest.value.as_slice().try_into().ok())
                .ok_or(ProvenanceError::InvalidDigest(component))?;
        }
        Ok(Self { digests })
    }
}

impl ProvenanceSnapshot {
    /// Return every changed component in stable field order; never truncate.
    pub fn differences(&self, other: &Self) -> Vec<ProvenanceComponent> {
        ProvenanceComponent::ALL
            .into_iter()
            .enumerate()
            .filter_map(|(index, component)| {
                (self.digests[index] != other.digests[index]).then_some(component)
            })
            .collect()
    }
}

/// Metadata freshness only; this value never grants access or factor admission.
#[derive(Clone, Debug, Eq, PartialEq)]
pub enum ProvenanceAssessment {
    /// All six fields match a service-resolved current context.
    Current,
    /// Valid historical evidence does not match the requested current context.
    Stale(Vec<ProvenanceComponent>),
    /// No authoritative current context was available.
    Unresolved,
}

impl ProvenanceAssessment {
    /// Refuse stale or unresolved metrics at a current-result consumption boundary.
    ///
    /// # Errors
    /// Returns `Stale` with every changed field, or `UnresolvedCurrent`.
    pub fn require_current(&self) -> Result<(), ProvenanceError> {
        match self {
            Self::Current => Ok(()),
            Self::Stale(changed) => Err(ProvenanceError::Stale(changed.clone())),
            Self::Unresolved => Err(ProvenanceError::UnresolvedCurrent),
        }
    }
}

/// Integrity and freshness failures, never deterministic factor rejection.
#[derive(Clone, Debug, Eq, PartialEq)]
pub enum ProvenanceError {
    /// A required digest is absent or does not contain exactly 32 bytes.
    InvalidDigest(ProvenanceComponent),
    /// A result does not describe the immutable inputs of its original run.
    RecordingMismatch(Vec<ProvenanceComponent>),
    /// Historical metrics cannot be used as current results.
    Stale(Vec<ProvenanceComponent>),
    /// The owner could not resolve the current computational inputs.
    UnresolvedCurrent,
}

impl fmt::Display for ProvenanceError {
    fn fmt(&self, formatter: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            Self::InvalidDigest(component) => {
                write!(
                    formatter,
                    "invalid provenance digest: {}",
                    component.as_str()
                )
            }
            Self::RecordingMismatch(changed) => {
                write!(
                    formatter,
                    "recorded provenance differs from frozen inputs: {changed:?}"
                )
            }
            Self::Stale(changed) => write!(formatter, "research metrics are stale: {changed:?}"),
            Self::UnresolvedCurrent => {
                formatter.write_str("current research provenance is unresolved")
            }
        }
    }
}

impl std::error::Error for ProvenanceError {}

/// Check original-run integrity first, then freshness against resolved inputs.
///
/// The same factor, backtest specification, sample and seed must be resolved by
/// the caller separately. Missing current context never implies freshness.
/// This pure operation performs no I/O and does not rewrite historical evidence.
///
/// # Errors
/// Returns `RecordingMismatch` even when the current context is unavailable or
/// happens to match the incorrectly recorded result.
pub fn assess_provenance(
    recorded: &ProvenanceSnapshot,
    frozen: &ProvenanceSnapshot,
    current: Option<&ProvenanceSnapshot>,
) -> Result<ProvenanceAssessment, ProvenanceError> {
    let mismatch = recorded.differences(frozen);
    if !mismatch.is_empty() {
        return Err(ProvenanceError::RecordingMismatch(mismatch));
    }
    let Some(current) = current else {
        return Ok(ProvenanceAssessment::Unresolved);
    };
    let changed = recorded.differences(current);
    if changed.is_empty() {
        Ok(ProvenanceAssessment::Current)
    } else {
        Ok(ProvenanceAssessment::Stale(changed))
    }
}
